Device model

remexre opened the pull request:
On top of https://github.com/UMN-Kernel-Object/ukoos/pull/63; merge that one first, then rebase this.

---

- Adds initializers. These are used to register drivers with the appropriate subsystems.
- Adds the device model and documentation about it.
- Adds support for enumerating devices in the Devicetree.
- Adds the UART device class.
- Adds an incomplete ns16550a driver. This is the UART we get in QEMU, and _most_ UARTs in the wild are compatible with it. This driver is currently very basic, since we don't have interrupts (#9) and it doesn't do DMA.

---

Relevant to #10, but it is **not** resolved by this PR.

The UART device class and ns16550a driver here are not very finished.
They currently act as an example of the device model, but don't implement all of what's needed for the driver itself.

remexre committed:
Adds initializers and documentation about them.

Signed-off-by: Amy Ringo <me@remexre.com>

remexre committed:
Makes devicetree_print prints the path to the node.

Signed-off-by: Amy Ringo <me@remexre.com>

remexre committed:
Adds the UART device class.

Signed-off-by: Amy Ringo <me@remexre.com>

remexre committed:
Adds the ns16550a driver.

Signed-off-by: Amy Ringo <me@remexre.com>

remexre committed:
Adds support for enumerating devices in a Devicetree.

Signed-off-by: Amy Ringo <me@remexre.com>

remexre committed:
Adds a notion of device objects.

Signed-off-by: Amy Ringo <me@remexre.com>

remexre committed:
Adds support for pointers to align_down and align_up.

Signed-off-by: Amy Ringo <me@remexre.com>

remexre committed:
Implements iomem_map and iomem_unmap.

These functions are for mapping IO memory.

Signed-off-by: Amy Ringo <me@remexre.com>

remexre commented:
Since there are other drivers in development (#81), despite the actual ns16550a driver and UART device class here being very unfinished, I'm un-drafting this and looking for review.

KieranMusser commented:
I'll look a bit more into this later (as I rewrite my driver), but should we include PCI somewhere in this? I feel it might be good to separate the PCI driver from the RTL8139 driver. I'll be doing all the work for PCI as a part of the RTL fixing, but PCI is useful to more than just that one driver.

remexre commented:
> I'll look a bit more into this later (as I rewrite my driver), but should we include PCI somewhere in this? I feel it might be good to separate the PCI driver from the RTL8139 driver.

Depends how big it is, I suppose; if the PCI and RTL8139 drivers are small enough together that they make sense to review together, then eh, whatever's fine.

If the PCI driver is large enough to need separate review, I would make a separate PR for it that has a target branch of this one, then set the RTL8139 PR to have a target branch of that one.

A few advantages there:

- The GitHub UI only shows the diff against that other branch, so your reviewers don't end up re-reviewing commits that're really from the other branch.
- The list of PRs makes it visually obvious when a branch depends on another in this way.
- Once the target branch gets merged, GitHub will email you and your reviewer(s). The message is phrased a bit rudely ("dismissed their stale review"), but... it's at least a ping lol.

jastintime commented on src/kernel/include/device.h:
```diff
@@ -0,0 +1,85 @@
+  struct list_head parent_children;
+
+  /**
+   * The children of the node.
```
this is an unnecessary comment.

remexre replied:
ratio

jastintime commented on src/kernel/include/device.h:
```diff
@@ -0,0 +1,85 @@
+ * Registers a device in the tree of devices and with its device class.
+ *
+ * - `name` must be filled out.
+ * - `parent` and `device_class` must be `nullptr`.
```
im confused, the attribute says nonnull but the comment says must be nullptr.

remexre replied:
Fixed in a9eab9aca7ecd9ad978ece213e689d9b21fb0c6c

jastintime commented on doc/kernel/devices.md:
```diff
@@ -0,0 +1,221 @@
+
+## Device classes
+
+Each device probably provides some functionality other than simply existing.
```
this is probably not necessary.

remexre replied:
fixed in 3ab30385781575e2698088f868204f305c85d8d2

jastintime commented on doc/kernel/devices.md:
```diff
@@ -0,0 +1,221 @@
+
+Typically, a device class consists of three things: a vtable struct, a device struct, and a list of devices.
+
+The vtable is simply a struct with methods for that kind of device.
```
c++ 😢

jastintime commented on src/kernel/arch/riscv64/paging.c:
```diff
@@ -205,9 +206,62 @@ bool mm_paging_unmap(uaddr va, paddr *pa) {
+
+  uaddr lo, hi;
+  struct vma *vma;
+  vma = vma_alloc(&kernel_virtual_allocator, size >> 12);
```
size >> 12 is some magic

remexre replied:
i'll do this tmrw

remexre replied:
fixed in d09e768a924430c45e74746f81e7b51936df7f4b

"tomorrow"

jastintime commented on src/kernel/arch/riscv64/paging.c:
```diff
@@ -205,9 +206,62 @@ bool mm_paging_unmap(uaddr va, paddr *pa) {
+    goto fail;
+  vma_bounds(vma, &lo, &hi);
+
+  usize i;
```
why is the usize outside of the for loop

remexre replied:
used in the second for loop

jastintime commented on src/kernel/arch/riscv64/paging.c:
```diff
@@ -205,9 +206,62 @@ bool mm_paging_unmap(uaddr va, paddr *pa) {
+  vma_bounds(vma, &lo, &hi);
+
+  usize i;
+  for (i = 0; i < hi - lo; i += 4096) {
```
4096 magic, #define the page size or constexpr

remexre replied:
fixed in d09e768a924430c45e74746f81e7b51936df7f4b

jastintime commented on src/kernel/arch/riscv64/paging.c:
```diff
@@ -205,9 +206,62 @@ bool mm_paging_unmap(uaddr va, paddr *pa) {
+
+  void *ptr_aligned;
+  ptr_aligned = align_down(ptr, 12);
+  size = align_up(size + (usize)(ptr - ptr_aligned), 12);
```
kieran said void ptr math might be UB

remexre replied:
gcc defines it: https://gcc.gnu.org/onlinedocs/gcc/Pointer-Arith.html

jastintime commented on src/kernel/devices/uart.c:
```diff
@@ -0,0 +1,9 @@
+/*
+ * SPDX-FileCopyrightText: 2025 ukoOS Contributors
```
https://hynek.me/til/copyright-years/ can we just get rid of copyright years, arguably not needed

remexre replied:
Resolved in 06b66739c941096f56daad31d3ed1623308bc21b.

jastintime commented on src/kernel/drivers/ns16550a.c:
```diff
@@ -0,0 +1,209 @@
+
+/**
+ * The bits of the Interrupt Enable Register.
+ */
```
can we add someone a link to the spec sheet for this UART device.

jastintime replied:
im unable to verify these cant find the spec sheet.

remexre replied:
added to website

jastintime commented on src/kernel/drivers/ns16550a.c:
```diff
@@ -0,0 +1,209 @@
+  return &device->device;
+
+fail:
+  if (device) {
```
might be a good idea to have a device_free(), so we don't have to remember to free the format name, unmap and free the whole thing.

remexre replied:
i think i was against this because it wouldn't actually run `free(device->device);`? like, in this code, you'd be doing:

```
if (device) {
    device_free(device->device);
    iomem_unmap(device->regs, reg_size);
}
free(device);
```

which is the same number of lines and seems less clear

jastintime replied:
true

jastintime commented on src/kernel/include/devices/uart.h:
```diff
@@ -0,0 +1,49 @@
+  const struct uart_ops *ops;
+
+  /**
+   * A pointer to the _same_ device.
```
what does _same_ device mean here.

remexre replied:
fixed in a18df5c2aa39fd79c21f81ce9651a89dfa10da29

jastintime commented on src/kernel/include/mm/paging.h:
```diff
@@ -74,4 +74,18 @@ bool mm_paging_map(uaddr va, paddr pa, enum page_permissions perms);
 
+/**
+ * Maps a region of memory for memory-mapped IO. This may involve special flags
+ * or other setup to prevent accesses from being cached.
```
where would this setup be? or do we not have that option yet?

remexre replied:
don't have the option yet; it's not needed on the duo s, but it would be the Svpbmt extn; this extension is part of RVA23 and the jupiter supports it too.

basically:

- the risc-v m-mode software is responsible for correctly configuring things behind the OS's back when you're mapping an io device
- if you have a PCI device that's reading from normal RAM, you need this to ensure that IO fences treat writes _to that normal RAM_ as IO accesses, to ensure they make it to RAM before the PCI device reads from it; this is what Svpbmt is for
  - i think it'd be reasonable for us to use Svpbmt when available (i'll file a ticket), and say "we don't support PCI on devices without Svpbmt"

remexre replied:
filed as https://github.com/UMN-Kernel-Object/ukoos/issues/93

jastintime commented on src/kernel/devicetree.c:
```diff
@@ -659,3 +688,122 @@ void devicetree_add_entropy(struct devicetree_node *root) {
+
+  if (!handlers) {
+    assert(!handlers_len);
+    handlers_cap = 16;
```
magic value

remexre replied:
arbitrary value; any non-zero constant here is correct

remexre committed:
Clarifies wording in docs.

Signed-off-by: Amy Ringo <me@remexre.com>

remexre committed:
*: remove copyright years

This aligns with the guidance in
https://www.linuxfoundation.org/blog/blog/copyright-notices-in-open-source-software-projects.

Signed-off-by: Amy Ringo <me@remexre.com>

remexre committed:
clarify preconditions in comment

Signed-off-by: Amy Ringo <me@remexre.com>

remexre committed:
clarify wording in comment

Signed-off-by: Amy Ringo <me@remexre.com>

remexre committed:
add READ_ONCE and WRITE_ONCE

Signed-off-by: Amy Ringo <me@remexre.com>

remexre committed:
add riscv constants header, use it in paging

Signed-off-by: Amy Ringo <me@remexre.com>

remexre committed:
ns16550a: use READ_ONCE and WRITE_ONCE in write_byte

Signed-off-by: Amy Ringo <me@remexre.com>

remexre commented:
I think I addressed everything? Ready for re-review @jastintime

jastintime reviewed (approved):
lgtm
